Add Performance Points & Star Rating Updates article (#6337)
ppy/osu-wiki: news/2021-11-08-performance-points-star-rating-updates.md

- [Panda Eyes & Teminite - Immortal Flame (feat. Anna Yvette) (Sotarks) [Eternal Blaze]](https://osu.ppy.sh/beatmapsets/703957#osu/1489207) 7.05* --> 6.94*

As for the formula change itself, you can compare and contrast between these:

Old formula: ![](/wiki/shared/news/2021-11-08-performance-points-star-rating-updates/old-sr-white.png)
New formula: ![](/wiki/shared/news/2021-11-08-performance-points-star-rating-updates/new-sr-white.png)

Interested in what each term means? [**This document**](https://docs.google.com/document/d/10DZGYYSsT_yjz2Mtp6yIJld0Rqx4E-vVHupCqiM4TNI/edit?usp=sharing) written up by [Naitsirk](https://osu.ppy.sh/users/8202998) goes into the nitty gritty details regarding the terms and the derivation itself, for those more mathematically inclined.

### New Flashlight difficulty skill

Flashlight in difficulty calculation is now revamped, as it moves from an object count-based multiplier applied to the Aim skill to a brand new skill of its own. 

The difficulty that Flashlight adds to a map is now measured much more accurately, thanks to [**MBmasher**](https://osu.ppy.sh/users/4498616)'s change in [this pull request](https://github.com/ppy/osu/pull/14217).

> Previously, the object count-based method posed a number of issues. Flow aim (used for hitting streams) is mostly represented by the speed skill, not aim. Additionally, the aim skill, unsurprisingly, only measures the aim difficulty, not the flashlight difficulty. To put this simply, there will be types of jumps that are hard to aim but quite easy to read with flashlight, and vice versa.

Reviewer: ```suggestion
Previously, the object count-based method posed a number of issues. Flow aim (used for hitting streams) is mostly represented by the speed skill, not aim. Additionally, the aim skill, unsurprisingly, only measures the aim difficulty, not the Flashlight difficulty. To put this simply, there will be types of jumps that are hard to aim but quite easy to read with Flashlight, and vice versa.
```

> For example, let’s consider a familiar type of jump pattern: a back and forth. This is an example of a jump that can have high aim difficulty, but relatively low flashlight difficulty. The distance between the first object and the third object in a back and forth is typically small enough for it to be within the flashlight radius, thus allowing a player to read the position of that circle very easily.

Reviewer: ```suggestion
For example, let's consider a familiar type of jump pattern: a back and forth. This is an example of a jump that can have high aim difficulty, but relatively low Flashlight difficulty. The distance between the first object and the third object in a back and forth is typically small enough for it to be within the Flashlight radius, thus allowing a player to read the position of that circle very easily.
```

![](/wiki/shared/news/2021-11-08-performance-points-star-rating-updates/visible-fl-area.png)

The creation of a Flashlight skill makes it possible to have it contribute to star rating. [This has been enabled](https://github.com/ppy/osu/pull/14753) following the results of the recent osu! performance points and star rating survey, which revealed that players desire star rating to be a metric of difficulty, and that they want subjective skills included. Ask and thou shall receive!

> It is also worth noting that the Hidden mod now applies a 1.3x multiplier on the Flashlight skill. The visibility of hitobjects are decreased significantly due to fading of circles and the absence of approach circles.

Reviewer: ```suggestion
It is also worth noting that the Hidden mod now applies a 1.3x multiplier on the Flashlight skill. The visibility of hit objects are decreased significantly due to the fading of circles and the absence of approach circles.
```

### Speed hard cap removal

The speed hard cap in osu! has been removed by [**emu1337**](https://osu.ppy.sh/users/2185987) and [**Apo11o**](https://osu.ppy.sh/users/9558549) in [this pull request](https://github.com/ppy/osu/pull/14617), allowing 300+ BPM streams to be weighed accurately. 

The speed cap was previously set to 300 BPM to prevent abuse cases, which means that calculating difficulty on beatmaps with higher BPM requires special consideration. Two adjustments were made to account for that:

- Speed is nerfed on doubles that are doubletappable. If the time since the last object (deltatime) for the current object is smaller than the 300 hit window, we instead interpolate between that and the deltatime for the previous object.
- Speed is nerfed if successive objects are faster than the 300 hit window. This is intended to target cheesing of faster bursts and streams if the hit window is way larger than the time between notes.

The intention behind this change is to accommodate for future improvement, as there are a handful of players who have proven themselves to be competent at such high speeds. 

> As an added bonus, the star ratings of absurdly fast maps are now measured more accurately, which means you can gawk as much as you’d like at the now even-higher star ratings on them.

Reviewer: ```suggestion
As an added bonus, the star ratings of absurdly fast maps are now measured more accurately, which means you can gawk as much as you'd like at the now even-higher star ratings on them.
```

### Minor changes

Alongside all those big changes above, there are some changes on the smaller side of things:

> - [Sliderbreaks are now approximated into the total miss count](https://github.com/ppy/osu/pull/15086) in performance calculation, following [**StanR**](https://osu.ppy.sh/users/7217455)’s pull request. This is guessed by comparing the achieved combo to the maximum combo possible on a map. 

Reviewer: ```suggestion
- [Sliderbreaks are now approximated into the total miss count](https://github.com/ppy/osu/pull/15086) in performance calculation, following [**StanR**](https://osu.ppy.sh/users/7217455)'s pull request. This is guessed by comparing the achieved combo to the maximum combo possible on a map. 
```

If you find yourself scratching your head when reading, consider consulting the [performance points](/wiki/Performance_points) and [star rating](/wiki/Beatmapping/Star_rating) wiki articles to gain a greater understanding of the topics.

We will be covering only changes made to the **osu!** game mode in this post.

## osu! changes

As a reminder of the various core values in the difficulty calculation algorithm:

- **Aim**, which considers how difficult it is to consistently hit notes back-to-back during play (affected by things like approach rate, jumps, Flashlight, Hidden, Hard Rock and more)
- **Speed**, which considers how quickly new elements are presented to the player during the course of the map (affected by things like the number of hit objects over time, BPM, Double Time and Half Time)
- **Accuracy**, which considers the player's ability to stay on time rhythmically and how difficult this is to achieve (affected by things like Hidden, Hard Rock, Flashlight and more)

> - **Strain**, which considers how often and how long a player has to endure intense or technically demanding patterns during play (affected by things like streams, fast jumps and more)

Reviewer 2: This is mostly an introductory recap. [Was present on past posts, too.](https://osu.ppy.sh/home/news/2021-07-27-performance-points-star-rating-updates)

> To help facilitate faster review and uptake of incoming changes, a Performance Points Committee has been formed, consisting of [**StanR**](https://osu.ppy.sh/users/7217455), [**emu1337**](https://osu.ppy.sh/users/2185987), [**MBmasher**](https://osu.ppy.sh/users/4498616) and [**Apo11o**](https://osu.ppy.sh/users/9558549).

Reviewer 2: The first public reference I can find about it is [this discord message](https://discord.com/channels/188630481301012481/380598781432823815/883167719942414366).

I believe the rationale for forming it is just to ensure consensus and coherent movement with the regard to how changes are made, to avoid back-and-forths.

With every large-scale change, there is always a chance that new issues may emerge as a result. As players and mappers adapt to these new changes, a counter-balancing patch may be needed in response. 

We intend to stay alert on these developments in the future and promptly address issues that arise in future updates. For this reason, we implore mappers and players to try leaning into these changes and have fun with them!

### Rhythm complexity

With [a new change](https://github.com/ppy/osu/pull/14395) submitted by [**Xexxar**](https://osu.ppy.sh/users/2773526), rhythm complexity will now play a more important role in determining the star rating of a map. 

This change introduces an algorithmic assessment of the complexity of note patterns and will introduce a more accurate assessment for maps that feature atypical combinations of triples, doubles, quints and much more!

Consider these two rhythms:

![](/wiki/shared/news/2021-11-08-performance-points-star-rating-updates/rhythm-complexity.png)

> Before this difficulty calculation change, the first pattern would have always had greater strain than the second.  In reality, the second can be as demanding or even harder than the first depending on the map. With these new changes, the algorithm now more accurately reflects this.

Reviewer: ```suggestion
Before this difficulty calculation change, the first pattern would have always had greater strain than the second. In reality, the second can be as demanding or even harder than the first depending on the map. With these new changes, the algorithm now more accurately reflects this.
```

- islands becoming smaller or larger (size 5 to size 3, for example)
- islands which change from even to odd sizes
- the ratio of time elapsed between islands (changes like 1/4 streams into 1/3 streams)

These changes all collectively combine to produce a rhythm value which then adjusts the difficulty of a pattern, from which star rating is ultimately derived.

We hope this change will bring a renewed interest in rhythmically complex songs on osu! and will incentivize players to improve their finger control. 

Good luck setting new, impressive top plays that showcase your skills!

### Total star rating adjustments

In a nutshell, osu! difficulty calculation processes a map and produces skill rating values. These skill values are then manipulated in two different ways to produce two different numbers that you’ll be familiar with: star rating and performance points. 

The correlation between these two numbers has now been improved, following the mathematical derivation work from [**Naitsirk**](https://osu.ppy.sh/users/8202998) in [this pull request](https://github.com/ppy/osu/pull/13986). 

> To summarize these changes briefly, maps which focus on one skill will now receive lower star rating than ones which focus on both. **These changes do not affect performance points.**

Reviewer 2: While true, this feels like it's beside the point of the change. The main point I'd emphasise here is that the old SR didn't really care about what the *composition* of a star rating was - as in how large the individual skill values are; it just summed the constituent parts and left it at that, while the new one combines them in a way analogous to how pp does. The statement that one-skill maps will be nerfed is a corollary of the above.

Here are a few examples:

- [DragonForce - Through the Fire and Flames (Ponoyoshi) [Myth]](https://osu.ppy.sh/beatmapsets/382400#osu/1001682) 6.1* -> 6.32*
- [Panda Eyes & Teminite - Immortal Flame (feat. Anna Yvette) (Sotarks) [Eternal Blaze]](https://osu.ppy.sh/beatmapsets/703957#osu/1489207) 7.05* -> 6.94*

As for the formula change itself, you can compare and contrast between these:

Old formula: ![](/wiki/shared/news/2021-11-08-performance-points-star-rating-updates/old-sr-white.png)

> New formula: ![](/wiki/shared/news/2021-11-08-performance-points-star-rating-updates/new-sr-white.png)

Reviewer 3: if the images are to be kept in the article they should be figures instead of inline. "\_ formula" can be the captions

> Interested in what each term means? [**This document**](https://docs.google.com/document/d/10DZGYYSsT_yjz2Mtp6yIJld0Rqx4E-vVHupCqiM4TNI/edit?usp=sharing) written up by [Naitsirk](https://osu.ppy.sh/users/8202998) goes into the nitty gritty details regarding the terms and the derivation itself, for those more mathematically inclined.

Reviewer 4: Also found this weird. If the formulas can be replaced with a layman's version that the average user might be able to parse then that would be best.

> Interested in what each term means? [**This document**](https://docs.google.com/document/d/10DZGYYSsT_yjz2Mtp6yIJld0Rqx4E-vVHupCqiM4TNI/edit?usp=sharing) written up by [Naitsirk](https://osu.ppy.sh/users/8202998) goes into the nitty gritty details regarding the terms and the derivation itself, for those more mathematically inclined.

Reviewer 2: Can only agree with the predecessors. I'd agree with snipping the formulas entirely. This plot from the google doc may get the point across better, I think (but only with explanation as to what is being shown):

![2021-11-08-082811_808x563_scrot](https://user-images.githubusercontent.com/20418176/140700850-ef05759c-be6e-404b-b27c-20dbb7cfbc0b.png)

I can theme it better using my jupyter notebooks I have stashed from the past newsposts if you think that'll work ok.

> Interested in what each term means? [**This document**](https://docs.google.com/document/d/10DZGYYSsT_yjz2Mtp6yIJld0Rqx4E-vVHupCqiM4TNI/edit?usp=sharing) written up by [Naitsirk](https://osu.ppy.sh/users/8202998) goes into the nitty gritty details regarding the terms and the derivation itself, for those more mathematically inclined.

Author: can try it if you want, if you think it's descriptive enough?

> Interested in what each term means? [**This document**](https://docs.google.com/document/d/10DZGYYSsT_yjz2Mtp6yIJld0Rqx4E-vVHupCqiM4TNI/edit?usp=sharing) written up by [Naitsirk](https://osu.ppy.sh/users/8202998) goes into the nitty gritty details regarding the terms and the derivation itself, for those more mathematically inclined.

Reviewer 2: Yeah I was just about to start preparing something to resolve my unresolved reviews. Sorry for the wait.

> Interested in what each term means? [**This document**](https://docs.google.com/document/d/10DZGYYSsT_yjz2Mtp6yIJld0Rqx4E-vVHupCqiM4TNI/edit?usp=sharing) written up by [Naitsirk](https://osu.ppy.sh/users/8202998) goes into the nitty gritty details regarding the terms and the derivation itself, for those more mathematically inclined.

Reviewer 2: Ok, so after some sanity loss due to failing to algebra properly, here's the best graph I can give for use instead of the formulas above. I'd just replace them wholesale with this graph.

![sr_pp_correlation](https://user-images.githubusercontent.com/20418176/140804434-87d9edb3-6c55-409b-a683-fd6408ea2cc5.png)

So the way this graph works, is that the axes are ratings for individual skills. The lines represent the combinations of aim & speed skills that a map needs to have to get a final rating of 7*. I don't know whether this is understandable, but I have no idea as to what could be better. Please advise whether you would like me to expand the description further.

Unfortunately I don't think using the ones from the original doc is viable as they are kind of confusing and arbitrary (the axes there don't really represent anything tangible). So I don't have any good idea how to graphically present the "more closely related to pp" part well.

I'd say the link to the doc can stay if someone _really_ wants to look at the mathematical derivation properly.

> Interested in what each term means? [**This document**](https://docs.google.com/document/d/10DZGYYSsT_yjz2Mtp6yIJld0Rqx4E-vVHupCqiM4TNI/edit?usp=sharing) written up by [Naitsirk](https://osu.ppy.sh/users/8202998) goes into the nitty gritty details regarding the terms and the derivation itself, for those more mathematically inclined.

Reviewer 4: New graph needs to be added to the article.

> Interested in what each term means? [**This document**](https://docs.google.com/document/d/10DZGYYSsT_yjz2Mtp6yIJld0Rqx4E-vVHupCqiM4TNI/edit?usp=sharing) written up by [Naitsirk](https://osu.ppy.sh/users/8202998) goes into the nitty gritty details regarding the terms and the derivation itself, for those more mathematically inclined.

Author: I think this does the trick far better than the formula ever did.

### New Flashlight difficulty skill

> Flashlight in difficulty calculation is now revamped, as it moves from an object count-based multiplier on the Aim skill to a brand new skill of its own. 

Reviewer 2: ```suggestion
Flashlight in difficulty calculation is now revamped, as it moves from an object count-based multiplier applied to the Aim skill to a brand new skill of its own. 
```

The difficulty that Flashlight adds to a map is now measured much more accurately, thanks to [**MBmasher**](https://osu.ppy.sh/users/4498616)’s change in [this pull request](https://github.com/ppy/osu/pull/14217).

> Previously, the object count-based method posed a number of issues. Flow aim (measured when playing streams) is mostly represented by the speed skill, not aim. Additionally, the aim skill (surprisingly) only measures the aim difficulty, not the flashlight difficulty. To put this simply, there will be types of jumps that are hard to aim and quite easy to read with flashlight, and vice versa.

Reviewer 2: ```suggestion
Previously, the object count-based method posed a number of issues. Flow aim (used for hitting streams) is mostly represented by the speed skill, not aim. Additionally, the aim skill, unsurprisingly, only measures the aim difficulty, not the flashlight difficulty. To put this simply, there will be types of jumps that are hard to aim but quite easy to read with flashlight, and vice versa.
```

The creation of a Flashlight skill makes it possible to have it contribute to Star Rating. [This has been enabled](https://github.com/ppy/osu/pull/14753) following the results of the osu! performance points and star rating survey, which revealed that players desire star rating to be a metric of difficulty, and that they want subjective skills included. Ask and thou shall receive!

It is also worth noting that the Hidden mod applies a 1.3x multiplier on the Flashlight skill. The visibility of hitobjects are decreased significantly due to fading of circles and the absence of approach circles.

Find yourself curious about the details? [MBmasher](https://osu.ppy.sh/users/4498616)’s write up about the implementation will very likely scratch that itch, so feel free to [**indulge in some Flashlight theory**](https://docs.google.com/document/d/11FBEEPXK8yI2d5uTZuebJC3S9nVUKgMD8WdvblERyYA/edit).

### Speed hard cap removal

The speed hard cap in osu! has been removed by [**emu1337**](https://osu.ppy.sh/users/2185987) and [**Apo11o**](https://osu.ppy.sh/users/9558549) in [this pull request](https://github.com/ppy/osu/pull/14617), allowing 300+ BPM streams to be weighed accurately. 

> The speed cap was previously set to 300 BPM to prevent abuse cases, which means that calculating higher BPMs requires consideration into those cases. Two factors are implemented with regard to this:

Reviewer 2: kinda reads weird. maybe this works better?

```suggestion
The speed cap was previously set to 300 BPM to prevent abuse cases, which means that calculating difficulty on beatmaps with higher BPM requires special consideration. Two adjustments were made to account for that:
```

> - Speed is nerfed on doubles that are doubletappable. If the time since the last object (deltatime) for the current object is smaller than the 300 hit window, we instead interpolate between that and the deltatime for the previous object.

Reviewer 2: I wouldn't bring "deltatime" into this at all, it's not easily understandable to readers. The interpolation detail also doesn't really matter. I would refer to hit windows instead as that seems more reasonable.

Author: attempted a reword to resolve/clarify this, not sure i have, am very open to suggestions on how to word this

> - Speed is nerfed on doubles that are double-tappable. If the time since the last object for the current object is smaller than the hit window for a 300, we instead take the difference between that object and the hit window of the last one.

Reviewer: Ah, no, I have no idea what this wording is trying to convey.

[The change in question for reference](https://github.com/ppy/osu/pull/14617/files#diff-2877a1e4555d4b301a40010c0693e8a8ce3f9bcd8759f1bc22b887d62a03b459R55-R56):

```cs
double speedWindowRatio = strainTime / greatWindowFull;

// Aim to nerf cheesy rhythms (Very fast consecutive doubles with large deltatimes between)
if (osuPrevious != null && strainTime < greatWindowFull && osuPrevious.StrainTime > strainTime)
                strainTime = Interpolation.Lerp(osuPrevious.StrainTime, strainTime, speedWindowRatio);
```

By inspection of the code, for a tight double, the time difference between its notes will be considered as if it was closer to the time difference of the gap before the double.

For a double with notes 25 ms apart (the minimum in calculations) and hit window of 50 ms (about AR 9.6 I think), the ratio of this weighted sum would be 0.5. For lower AR, doubles are nerfed harder and vice versa.

```suggestion
- Speed is nerfed on doubles that are double-tappable. If the time gap between two notes in a double is smaller than the hit window for a 300, the speed skill calculation will consider it as if it was closer to the time gap before the double.
```

... or whatever you find most illustrative.

Reviewer 2: The above seems fine, but I'd probably just omit/skip over the interpolation details, mostly. I'm not sure it's worth the attention. I'd probably leave it at something like

```suggestion
- A double can be considered double-tappable if the time interval between both notes of the double is small enough that their hit windows overlap, which makes it harder to miss both when a key is pressed twice in the overlap period. Those sorts of patterns are nerfed by artificially increasing the time gap between them.
```

Reviewer: I can agree with something like this, but I want to turn it around to front-load the information

```suggestion
Quick doubles with overlapping hit windows are nerfed. With a small enough time interval between two notes, they can still be perfectly hit by simply double-tapping during the overlap of their hit windows. This is now accounted for in calculations by artificially increasing the time gap between the notes.
```

> - Speed is nerfed if successive objects are faster than the 300 hit window. This is intended to target cheesing of faster bursts and streams if the hit window is way larger than the time between notes.

Reviewer: Would like this addressed. Main concern I guess is that "speed is nerfed" doesn't make much sense as a statement. "Speed pp" would also be less to-the-point. Can be a separate point still, but it could use some less redundant wording.

> - Speed is nerfed if successive objects are faster than the 300 hit window. This is intended to target cheesing of faster bursts and streams if the hit window is way larger than the time between notes.

Author: fair point, was holding off on this since I'm not 100% sure on either wording, but we can go with yours for the time being

The intention behind this change is to accommodate for future improvement, as there are a handful of players who have proven themselves to be competent at such high speeds. 

As an added bonus, the star ratings of absurdly fast maps are now measured more accurately, which means you can gawk as much as you'd like at the now even-higher star ratings on them.

### Minor changes

Alongside all those big changes above, there are some changes on the smaller side of things:

- [Sliderbreaks are now approximated into the total miss count](https://github.com/ppy/osu/pull/15086) in performance calculation, following [**StanR**](https://osu.ppy.sh/users/7217455)'s pull request. This is guessed by comparing the achieved combo to the maximum combo possible on a map. 
- [Instant spinners no longer give an insane amount of strain](https://github.com/ppy/osu/pull/14217), thanks to [**StanR**](https://osu.ppy.sh/users/7217455). This fixes maps [such as this one](https://osu.ppy.sh/beatmapsets/814850#osu/1901200), which abused this. 

> - Joz spotted that the initial strain for skills started at 1, not 0. [This has been fixed](https://github.com/ppy/osu/pull/15014).

Reviewer 5: ```suggestion
- [**Joz**](https://osu.ppy.sh/users/10644596) spotted that the initial strain for skills started at 1, not 0. [This has been fixed](https://github.com/ppy/osu/pull/15014).
```

- In preparation for osu!lazer, difficulty calculation for [the Blinds mod](https://github.com/ppy/osu/pull/14921) and [the Relax mod](https://github.com/ppy/osu/pull/14942) has been added by [**Apo11o**](https://osu.ppy.sh/users/9558549).
  - These could be potentially enabled in the future once osu!lazer leaderboard submissions go live, since the concept of "ranked mods" will be no more.

---

> Massive thanks to the star rating developers for all the changes above, the community members for providing continuous feedback, plus [Apo11o](https://osu.ppy.sh/users/9558549) and [Xexxar](https://osu.ppy.sh/users/2773526) for writing this post.
> 
> —osu!team

Reviewer 3: ```suggestion
Massive thanks to the star rating developers for all the changes above, and the community members for providing continuous feedback.

—Apo11o and Xexxar
```

Reviewer 2: These posts are generally always credited to `osu!team` anyhow.

Reviewer 4: Yeah, if anything the previous version read better (leaving `osu!team` as the sign-off). Xexxar is already mentioned above so could probably be left to just Apo11o, or omit completely as this post is going to be a community effort once all these reviews are applied.

Reviewer 3: I commented to point out that it literally says who wrote the post, followed by a sign-off that is different from who wrote the post. I don't have an opinion who counts as the author, just that it shouldn't contradict itself

> Massive thanks to the star rating developers for all the changes above, the community members for providing continuous feedback, and [Apo11o](https://osu.ppy.sh/users/9558549), [Xexxar](https://osu.ppy.sh/users/2773526) and [bastoo0](https://osu.ppy.sh/users/4864877) for writing this post.

Reviewer 6: May wanna remove bastoo0 since his contributions were related to the osu!catch section.

---
layout: post
title: "Performance Points & Star Rating Updates"
date: 2021-11-08 10:00:00 +0000
---

![](/wiki/shared/news/2021-07-27-performance-points-star-rating-updates/star-rating.jpg)

> Thanks to the efforts of various star rating contributors, we are excited to announce that a new series of difficulty calculation changes will be going live shortly. This newspost will go into the what, the who, the when and the how, so read on to find out more!

Reviewer 3: also for "star rating developers" on L130

> Thanks to the efforts of various star rating contributors, we are excited to announce that a new series of difficulty calculation changes will be going live shortly. This newspost will go into the what, the who, the when and the how, so read on to find out more!

Reviewer 4: I'd just use the term "community contributors" or "developers in the community" or something like that.
